framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,656 @@
+            AnnotationFieldType aft = a.def().fieldTypes.get(f.getKey());
+            sj.add(f.getKey() + "=" + IndexFileWriter.formatAnnotationValue(aft, f.getValue()));
+        }
+        return sj.toString();
+    }
+
+    /**
+     * Returns all annotations in {@code annos} in a form suitable to be printed as Java source
+     * code.
+     *
+     * <p>Each annotation is followed by a space, to separate it from following Java code.
+     *
+     * @param annos the annotations to format
+     * @return all annotations in {@code annos}, each followed by a space, in a form suitable to be
+     *     printed as Java source code
+     */
+    private static String formatAnnotations(Collection<? extends Annotation> annos) {
+        StringBuilder sb = new StringBuilder();
+        for (Annotation tla : annos) {
+            if (!isInternalJDKAnnotation(tla.def.name)) {
+                sb.append(formatAnnotation(tla));
+                sb.append(" ");
+            }
+        }
+        return sb.toString();
+    }
+
+    /**
+     * Formats the component types of an array via recursive descent through the array's scene-lib
+     * structure.
+     *
+     * @param e the array's scenelib type element
+     * @param arrayType the string representation of the array's type
+     * @return the type formatted to be written to Java source code, followed by a space character
+     */
+    private static String formatArrayType(ATypeElement e, String arrayType) {
+        StringBuilder result = new StringBuilder();
+        // If there are not annotations on the component, this will be true - unless there
+        // are also no annotations on the first array level. See comment below for how that
+        // case is handled.
+        int componentEndPos;
+        {
+            componentEndPos = arrayType.indexOf(' ');
+            if (arrayType.startsWith("@")) {
+                // While there are more explicit annotations, go to the next space.
+                while (arrayType.charAt(componentEndPos + 1) == '@') {
+                    componentEndPos = arrayType.indexOf(' ', componentEndPos + 1);
+                }
+                // Once all annotations have been skipped, we still have to skip to the end of the
+                // component type itself.
+                componentEndPos = arrayType.indexOf(' ', componentEndPos + 1);
+            }
+            // If the first array level doesn't have an annotation, then the spacing will be off,
+            // so use the position of the first '[' instead (which must be correct, since the
+            // first array doesn't have an annotation).
+            int firstArrayPos = arrayType.indexOf('[');
+            if (firstArrayPos < componentEndPos || componentEndPos == -1) {
+                componentEndPos = firstArrayPos - 1;
+            }
+        }
+
+        String componentType = arrayType.substring(0, componentEndPos + 1);
```

Comment:
It looks like either the first one of these ends with a space or the second one starts with a space.  (I'm not sure which.) 
 Does it matter which?  Is `formatArrayTypeImpl` robust against spaces in those locations?